### DISCUSSION TITLE
PDF settings Landscape property seems to work like it was named Portrait

### DISCUSSION THREAD

**Discussion by adellieux** [cefsharp/CefSharp#4447] [Q&A] [answered]
Hi,

I'm using CefSharp 109.1.110, under Windows 10 (developer laptop) and (Windows server 2012 R2, Windows 2019, Windows server 2016 : clients servers)

Compilation in x86 mode, and making a .Net Framework 4.6.1 WinForm activeX for old Visual FoxPro 9 applications.

I encounter a strange behaviour using Pdf print settings.
The Landscape property seems to work like it was named Portrait.

If i omit Landscape in the settings, or set it to false, then the generated PDF is in landsape orientation.
If i set Landscape to true, then the generated PDF is in portrait orientation.

This occurs with Cefsharp Minimal example repository, 109.1.110 too.
Just copy - pasted the following code, and called it from the "Go" buttonclick, just after the LoadUrl() call :             SavePdf("d:\\_test.pdf", "header text");

I may do something wrong.
Anybody having such  behaviour ?

Note that clients technical recommendations does not allow me to us more recent versions, because 109.1.110 is the last version to support Windows 7 / 8 / 8.1 and servers equivalents.

```
        private void SavePdf(string fileName, string header, string footer = " ")
        {
            if (!IsBrowserInitialized())
            {
                return;
            }
            _ = PrintToPdfTask(fileName, header, footer).Result;
        }

        private Task<bool> PrintToPdfTask(string fileName, string header, string footer)
        {
            var settings = new PdfPrintSettings
            {
                PrintBackground = false,

                DisplayHeaderFooter = true,
                HeaderTemplate = !string.IsNullOrWhiteSpace(header) ? $"<div style='margin:1px; width:100%; font-size:12px; text-align:center'>{header}</div>" : " ",
                FooterTemplate = !string.IsNullOrWhiteSpace(footer) ? $"<div style='margin:1px; width:100%; font-size:12px; text-align:center'>{footer}</div>" : " ",

                Scale = 1.0,

                Landscape = true,

                MarginType = CefPdfPrintMarginType.Custom,
                MarginTop = 0.4,
                MarginBottom = 0.4,
                MarginLeft = 0.4,
                MarginRight = 0.4,

                PaperHeight = 8.3,
                PaperWidth = 11.7,
            };

            return _browser.PrintToPdfAsync(fileName, settings);
        }

```

**Comment by amaitland** [comment, marked as answer]
> The Landscape property seems to work like it was named Portrait.

`CefSharp` does little more than pass the values to the underlying `CEF method`. So I suspect this is a problem in [CEF](https://github.com/chromiumembedded/cef)

https://github.com/cefsharp/CefSharp/blob/3a3aed61fa95e9dc61205a8e62180d3eb8ae4a82/CefSharp.Core.Runtime/Internals/CefBrowserHostWrapper.cpp#L90

https://github.com/chromiumembedded/cef/blob/c83b3cda24e2960141dc13e1f089edb93c939add/include/internal/cef_types.h#L2505

If the problem reproduces in the current supported version then it's probably worth stepping through with the debugger to confirm the correct value is being passed through. On the surface it looks like an upstream issue in CEF.

> Note that clients technical recommendations does not allow me to us more recent versions, because 109.1.110 is the last version to support Windows 7 / 8 / 8.1 and servers equivalents.

Unfortunately version 109 is no longer supported upstream so we won't be getting any more fixes.

**Comment by adellieux** [comment]
Hi @amaitland ,

back to work on CefSharp version 114.2.100
i made a new, little test program in C#, .Net Framework 4.6.2, 64 bits, manifest file.
i added two buttons to call chromiumWebBrowser1.PrintToPdfAsync() and generate two pdf files, based on www.google.fr, one in portrait mode, the other in landscape mode.
like in version 109, portrait and landscape are still behaving incorrectly.
as you said, it surely is a bug in CEF.
unfortunately, i'm not able to debug C++ (i have no knowledge of C++)

here is the code i used :

```
using CefSharp;
using CefSharp.WinForms;
using System;
using System.IO;
using System.Reflection;
using System.Windows.Forms;

namespace WindowsFormsApp1
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();

            Assembly currentAssembly = Assembly.GetExecutingAssembly();
            string assemblyPath = Path.GetDirectoryName(currentAssembly.Location);

            CefSettings settings = new CefSettings
            {
                BrowserSubprocessPath = Path.Combine(assemblyPath, "CefSharp.BrowserSubprocess.exe"),
                LocalesDirPath = Path.Combine(assemblyPath, "locales"),
                ResourcesDirPath = assemblyPath,
                CachePath = Path.Combine(Environment.GetEnvironmentVariable("TEMP", EnvironmentVariableTarget.User), "CefSharp"),
                Locale = "fr",
                RemoteDebuggingPort = 8088,
                LogSeverity = LogSeverity.Error,
                AcceptLanguageList = "fr-FR,fr;en-US,en"
            };

            Cef.EnableWaitForBrowsersToClose();
            Cef.Initialize(settings);

            chromiumWebBrowser1.LoadUrl("https://www.google.fr");
        }

        private void Form1_FormClosed(object sender, FormClosedEventArgs e)
        {
            Cef.WaitForBrowsersToClose();
            Cef.Shutdown();
        }

        private void button1_Click(object sender, EventArgs e)
        {
            CreatePdfFile("portrait");
        }

        private void button2_Click(object sender, EventArgs e)
        {
            CreatePdfFile("landscape");
        }

        private void CreatePdfFile(string mode)
        {
            var pdfsettings = new PdfPrintSettings
            {
                // On n'imprime pas le fond
                PrintBackground = false,

                // Afficher entêtes et bas de pages (revoir complètement pour gérer date, titre, pagination)
                DisplayHeaderFooter = true,
                HeaderTemplate = "<div style='margin:1px; width:100%; font-size:12px; text-align:center'>header</div>",
                FooterTemplate = "<div style='margin:1px; width:100%; font-size:12px; text-align:center'>footer</div>",

                // Echelle à 100%
                Scale = 1.0,

                // Portrait / Landscape : there seems to be a bug here, values seem to be inverted
                Landscape = string.Compare(mode, "landscape") == 0,

                // Marges = 1 cm
                MarginType = CefPdfPrintMarginType.Custom,
                MarginTop = 0.4,
                MarginBottom = 0.4,
                MarginLeft = 0.4,
                MarginRight = 0.4,

                // Dimensions papier en pouces = 21 x 29,7
                PaperHeight = 8.3,
                PaperWidth = 11.7,
            };

            chromiumWebBrowser1.PrintToPdfAsync($"d:\\{mode}.pdf", pdfsettings);
        }
    }
}

```

**Reply by amaitland** [reply]
Based on the code example this looks to be a mistake in your code.

```c#
PaperHeight = 8.3,
PaperWidth = 11.7,
```

**Reply by adellieux** [reply]
my bad...
it was there all along...
this sample code is based on our main code, and those lines of code are the result of a code refactoring.
both properties have been inverted, but not the values.
so, as far as i can understand, that means that CEF does not only use Landscape property, but uses those page values too, to try to autodetermine the page orientation ?
anyway, @amaitland , many thanks for pointing this out
